### ISSUE TITLE
Beacon node throws "could not find block in DB" error almost every epoch.

### ISSUE THREAD

**Issue by e6a6** [prysmaticlabs/prysm#10156]
# 🐞 Bug Report

### Description
I recently missed a block proposal 4 times in a row. Avg. Effectiveness is close to 100%, and attestation has no problems. Only the block proposal was missed. There were no issues until Nov 2021. However, from Dec 2021, only block proposals started to be missed.

At first I thought I was simply out of luck or just had temporary network issue. Because I haven't changed any settings from the beginning of 2021 until now, except for updates of geth and prysm. However, when Jan 2022, another block proposal was missed. At this point, I decided that something was wrong. I checked the log to determine the cause. And, I saw that a huge amount of the same error in the beacon node. The same error is repeated 10 or more times simultaneously per epoch, and this error occurs almost every epoch.

<pre><code>
level=error msg="Could not retrieve pre state" error="could not get pre state for epoch xxxxx: could not get state summary: could not find block in DB" prefix=sync
<i>(where xxxxx is the latest epoch number.)</i>
</code></pre>

I think that this error was the root cause of the block proposal miss, and I tried the following to solve this error.

 - Restart the virtual machine
 - Restart the host machine
 - Remove prysm DB, and sync again from scratch
 - Use infura instead of local geth for ETH 1 chain connection
 - Test RAM errors with MemTest86
 - SSD S.M.A.R.T. Information check
 - Using an old version prysm (v2.0.4, v2.0.3)

But I kept getting the same error, and I guessed it was an unknown hardware problem. So I bought new hardware and set up again the node from scratch. It used more powerful hardware than before. CPU is Intel i7-1165G7, RAM is 64GB, SSD is NVMe 4TB. For personal reasons, I virtualized Ubuntu 20.04 using Windows 10 Pro and Hyper-V. 

I waited for the chain to completely sync by running only the beacon node. I checked the logs for several hours after the sync was complete. The "Could not retrieve pre state" error no longer occurred. However, not long after running the validator I met the same error again in the beacon.

<pre><code>
Jan 31 00:31:38 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:31:38" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95658: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:31:38 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:31:38" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95658: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:31:38 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:31:38" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95658: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:38:03 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:38:03" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95659: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:38:03 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:38:03" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95659: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:38:03 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:38:03" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95659: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:38:03 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:38:03" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95659: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
Jan 31 00:50:48 ETH-Virtual-Machine prysm.sh[15984]: time="2022-01-31 00:50:48" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95661: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

I am still getting the same error. Attestation is fine, but expect to miss out on block proposals again someday. Can you tell me how to solve this error?

### Has this worked before in a previous version?

I tested by downgrading to v2.0.4 and v2.0.3 but got the same error. Prior to November 2021, there were no issues with any version. 

## 🔬 Minimal Reproduction

I used systemd service to execute prysm (beacon, validator). An error occurs when the beacon is synced and the validator starts running. The following is the content of the systemd service file I used.

<pre><code>
[Unit]
Description=Beacon
After=geth.target

[Service]
Type=simple
User=beacon
ExecStart=/home/eth2/beacon/bin/prysm.sh beacon-chain \
     --datadir=/home/eth2/beacon \
     --http-web3provider=http://10.8.0.2:8545 \
     --fallback-web3provider=https://mainnet.infura.io/v3/xxxxxxxxxxxxxxxxxx \
     --p2p-max-peers=70 \
     --slasher \
     --accept-terms-of-use

KillMode=process
KillSignal=SIGINT
TimeoutStopSec=90

Restart=always
RestartSec=5

[Install]
WantedBy=multi-user.target
</pre></code>

<pre><code>
[Unit]
Description=Validator
After=beacon.target

[Service]
Type=simple
User=validator
ExecStart=/home/eth2/validator/bin/prysm.sh validator \
  --datadir=/home/eth2/validator  \
  --wallet-dir=/home/eth2/validator/wallet/prysm-wallet \
  --wallet-password-file=/home/eth2/validator/wallet/password \
  --accept-terms-of-use

Restart=always
RestartSec=5

[Install]
WantedBy=multi-user.target
</pre></code>

## 🔥 Error
beacon node
<pre><code>
Jan 23 15:34:00 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:34:00" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 93999: could not get state summary: could not find block in DB" prefix=sync
Jan 23 15:34:00 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:34:00" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 93999: could not get state summary: could not find block in DB" prefix=sync
Jan 23 15:46:01 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:46:01" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 287822 but received 287592" prefix="rpc/validator"
Jan 23 15:53:14 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:53:14" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 94002: could not get state summary: could not find block in DB" prefix=sync
Jan 23 15:53:14 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:53:14" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 94002: could not get state summary: could not find block in DB" prefix=sync
</code></pre>

validator (the moment missed the block proposal)
<pre><code>
Jan 23 15:46:11 ETH-Virtual-Machine prysm.sh[5822]: time="2022-01-23 15:46:11" level=error msg="Failed to request block from beacon node" blockSlot=3008028 error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" prefix=validator pubKey=0xa05197a22ffb
</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Ubuntu 20.04 with Hyper-V virtualization on Windows 10 Pro.
And, running geth and prysm (beacon, validator) in the same Ubuntu VM.
(Host machine CPU: Intel i5-10210U, RAM 64GB (allocated 40GB to VM), SSD: 4TB (allocated all to VM))

A few days ago I replaced the hardware.

Ubuntu 20.04 with Hyper-V virtualization on Windows 10 Pro.
I separated the VMs for geth and prysm.
(Host machine CPU: Intel i7-1165G7, RAM 64GB (allocated 20GB to geth, 16GB to prysm), SSD: NVMe 4TB (allocated 1.5TB to prysm))
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
v2.0.5
I tried to change to version 2.0.4 and 2.0.3, but the same error occurred. 
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

**Comment by nisdas** [comment]
Hey @e6a6 , 

Thank you for the report. This is an interesting error as you still get the same error despite a resync. 
This log maybe interesting to us and related to #10120 :
```
Jan 23 15:46:01 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:46:01" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 287822 but received 287592" prefix="rpc/validator"
```
cc @rkapka 

However, the other errors you have are pretty unusual and the consistency of them might indicate some sort of hardware issue. You mentioned you got new hardware to what you had earlier, which parts of your machine did you exactly change and which other parts have you kept the same ?

For now, you can restart your node and it should fix the `Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 287822 but received 287592` . Let us know if you still have any missed proposals after that.

**Comment by e6a6** [comment]
Thanks for the response. First post was not clear, so I describe my hardware and the flow along the time series in more detail.

<pre><code>
<b>[Machine A]</b>
16 Jan 2021 ~ 31 Jan 2022
Model: Intel NUC10i5FNH (CPU: i5-10210U, RAM: 64GB, SSD: Samsung 860 EVO 4TB)
Ubuntu 20.04 with Hyper-V virtualization on Windows 10 Pro.
And, running geth and prysm (beacon, validator) in the same Ubuntu VM.
</pre></code>

I started staking in January 2021 using Machine A, and did not experience any issues through November. (include Nov) However, from December 2021, only block proposals started to be missed. There were no changes to the hardware and surroundings. Software changes were only updates of geth and prysm.

From Dec 22, 2021 to Jan 23, 2022, there were a total of 4 block proposals and all were missed. The following is the error message of validator and beacon when block proposal is missed 4 times. Since there are too many identical "could not find block in DB" errors in beacon, I've only imported some.

<pre><code>
<b>[Validator]</b>
(2021) Dec 22 07:11:11 ETH-Virtual-Machine prysm.sh[15852]: time="2021-12-22 07:11:11" level=error msg="Failed to request block from beacon node" blockSlot=2775053 error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" prefix=validator pubKey=0xb729e6a3fdaf

<b>[Beacon]</b>
Dec 22 07:08:25 ETH-Virtual-Machine prysm.sh[15817]: time="2021-12-22 07:08:25" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 86720: could not get state summary: could not find block in DB" prefix=sync
Dec 22 07:11:02 ETH-Virtual-Machine prysm.sh[15817]: time="2021-12-22 07:11:02" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 275318 but received 275089" prefix="rpc/validator"
Dec 22 07:21:14 ETH-Virtual-Machine prysm.sh[15817]: time="2021-12-22 07:21:14" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 86722: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

<pre><code>
<b>[Validator]</b>
(2021) Dec 27 20:56:35 ETH-Virtual-Machine prysm.sh[1045]: time="2021-12-27 20:56:35" level=error msg="Failed to request block from beacon node" blockSlot=2815180 error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" prefix=validator pubKey=0x85bf10fde124

<b>[Beacon]</b>
Dec 27 20:54:01 ETH-Virtual-Machine prysm.sh[1044]: time="2021-12-27 20:54:01" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 87974: could not get state summary: could not find block in DB" prefix=sync
Dec 27 20:56:27 ETH-Virtual-Machine prysm.sh[1044]: time="2021-12-27 20:56:27" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 276988 but received 276951" prefix="rpc/validator"
Dec 27 21:06:49 ETH-Virtual-Machine prysm.sh[1044]: time="2021-12-27 21:06:49" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 87976: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

<pre><code>
<b>[Validator]</b>
(2022) Jan 12 14:25:35 ETH-Virtual-Machine prysm.sh[1052]: time="2022-01-12 14:25:35" level=error msg="Failed to request block from beacon node" blockSlot=2928425 error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" prefix=validator pubKey=0x85bf10fde124

<b>[Beacon]</b>
Jan 12 14:23:36 ETH-Virtual-Machine prysm.sh[1051]: time="2022-01-12 14:23:36" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 91513: could not get state summary: could not find block in DB" prefix=sync
Jan 12 14:25:26 ETH-Virtual-Machine prysm.sh[1051]: time="2022-01-12 14:25:26" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 283691 but received 283392" prefix="rpc/validator"
Jan 12 14:30:01 ETH-Virtual-Machine prysm.sh[1051]: time="2022-01-12 14:30:01" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 91514: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

At this time (about 17~21 Jan), I figured something was wrong and took action (mentioned in the first post) including resynchronizing the beacons of Machine A. But the problem is not solved and after a few days I get the same error below again.

<pre><code>
<b>[Validator]</b>
(2022) Jan 23 15:46:11 ETH-Virtual-Machine prysm.sh[5822]: time="2022-01-23 15:46:11" level=error msg="Failed to request block from beacon node" blockSlot=3008028 error="rpc error: code = DeadlineExceeded desc = context deadline exceeded" prefix=validator pubKey=0xa05197a22ffb

<b>[Beacon]</b>
Jan 23 15:34:00 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:34:00" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 93999: could not get state summary: could not find block in DB" prefix=sync
Jan 23 15:46:01 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:46:01" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 287822 but received 287592" prefix="rpc/validator"
Jan 23 15:53:14 ETH-Virtual-Machine prysm.sh[5791]: time="2022-01-23 15:53:14" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 94002: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

I decided that there must be an unknown hardware problem as well as software because I got the same error consistently. I built a whole new machine B to remove this uncertainty. I bought a separate NUC and also bought new parts. None of them were recycled from the old Machine A. The software was also newly downloaded and installed from the official website.

<pre><code>
<b>[Machine B]</b>
31 Jan 2022 ~
Model: Intel NUC11PAHi7 (CPU: i7-1165G7, RAM: 64GB, SSD: Seagate FireCuda 530 4TB)
Ubuntu 20.04 with Hyper-V virtualization on Windows 10 Pro.
I separated the VMs for geth and prysm.
</pre></code>

On January 31st, Machine B was all set up, including syncing the beacons. I took the old Machine A completely offline and removed it from power. Then, after connecting Machine B and importing the key, the validator was operated. However, the same error occurred again.

<pre><code>
Jan 31 16:12:26 ETH2-VM prysm.sh[5697]: time="2022-01-31 16:12:26" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95805: could not get state summary: could not find block in DB" prefix=sync
Jan 31 16:12:26 ETH2-VM prysm.sh[5697]: time="2022-01-31 16:12:26" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95805: could not get state summary: could not find block in DB" prefix=sync
Jan 31 16:12:26 ETH2-VM prysm.sh[5697]: time="2022-01-31 16:12:26" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95805: could not get state summary: could not find block in DB" prefix=sync
Jan 31 16:12:26 ETH2-VM prysm.sh[5697]: time="2022-01-31 16:12:26" level=error msg="Could not retrieve pre state" error="could not get pre state for epoch 95805: could not get state summary: could not find block in DB" prefix=sync
</pre></code>

As mentioned earlier, **this error occurs almost every epoch**. Normally, no other error occurs other than this. However, looking at past cases, the warning "Cached deposit trie is invalid, rebuilding it now: wanted the canonical count" seems to occur only when a block proposal is ahead. Are there actions non-experts can take to resolve the error? Or maybe there is a compatibility issue between Hyper-V and prysm?

**Comment by nisdas** [comment]
Thanks for the detailed information on this. Given that this is showing up on 2 different sets of hardware builds it is unlikely hardware is the issue. I am not aware of any incompatibility with Hyper-V and prysm but cannot rule that out at this stage. Given that you are running windows 10, do you know if your node runs without issues on windows ? You can try copying over the file to your windows partition(although a manual copy requires the node to be stopped) Also are you aware of any big changes to windows in november/december ? Or any related update in that time period.

**Comment by prestonvanloon** [comment]
Any chance you could upload the beaconchain.db? We could inspect to see if there is some corrupt data there.

**Comment by e6a6** [comment]
I use StopUpdates10 to completely block automatic updates of Windows 10. I do updates by manually disabling StopUpdates10 aperiodically. My memory isn't clear, but I didn't do any windows updates in Nov~Dec 2021, and even now (Machine A). However, because Machine B was built at the end of January 2022, the latest windows updates are applied.

Just tested with windows-amd64 binary of prysm 2.0.6 in native windows 10 environment as per your advice. In Machine A, the windows environment of that time was still preserved. So, I decided that it was suitable for testing. Of course the Hyper-V VM was already deleted a few days ago in the machine A. (to prevent slashing) 

I copied the entire beacon datadir from Machine B's virtual machine to Machine A's windows 10. And, created a batch file as follows and excute the beacon. It synced with no problems. After turn off Machine B, I imported the wallet to Machine A and excute the validator.

<pre><code>
beacon206 --datadir=d:\beacon --http-web3provider=https://mainnet.infura.io/v3/xxxxxxxxxxx --slasher --p2p-max-peers=70 --log-file d:\beacon.log --accept-terms-of-use

validator206 --datadir=d:\validator --wallet-dir=d:\validator\wallet\prysm-wallet --log-file d:\validator.log --accept-terms-of-use
</pre></code>

Attestations were fine, but after a while I get the same "could not find block in DB" error. Here is the beacon's error log. The detailed message is different from Ubuntu, but it looks like the same error. Likewise, this error occurs almost every epoch, and I only brought one because the same error occurs too many times.

<pre><code>
[2022-02-02 04:02:49] ERROR sync: Could not retrieve pre state error=could not find block in DB
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).RecoverStateSummary
	beacon-chain/state/stategen/getter.go:152
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).stateSummary
	beacon-chain/state/stategen/getter.go:134
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).loadStateByRoot
	beacon-chain/state/stategen/getter.go:180
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).StateByRoot
	beacon-chain/state/stategen/getter.go:58
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).getAttPreState
	beacon-chain/blockchain/process_attestation_helpers.go:36
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).AttestationTargetState
	beacon-chain/blockchain/receive_attestation.go:61
github.com/prysmaticlabs/prysm/beacon-chain/sync.(*Service).validateCommitteeIndexBeaconAttestation.func1
	beacon-chain/sync/validate_beacon_attestation.go:95
runtime.goexit
	src/runtime/asm_amd64.s:1371
could not get state summary
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).loadStateByRoot
	beacon-chain/state/stategen/getter.go:182
github.com/prysmaticlabs/prysm/beacon-chain/state/stategen.(*State).StateByRoot
	beacon-chain/state/stategen/getter.go:58
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).getAttPreState
	beacon-chain/blockchain/process_attestation_helpers.go:36
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).AttestationTargetState
	beacon-chain/blockchain/receive_attestation.go:61
github.com/prysmaticlabs/prysm/beacon-chain/sync.(*Service).validateCommitteeIndexBeaconAttestation.func1
	beacon-chain/sync/validate_beacon_attestation.go:95
runtime.goexit
	src/runtime/asm_amd64.s:1371
could not get pre state for epoch 96141
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).getAttPreState
	beacon-chain/blockchain/process_attestation_helpers.go:38
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).AttestationTargetState
	beacon-chain/blockchain/receive_attestation.go:61
github.com/prysmaticlabs/prysm/beacon-chain/sync.(*Service).validateCommitteeIndexBeaconAttestation.func1
	beacon-chain/sync/validate_beacon_attestation.go:95
runtime.goexit
	src/runtime/asm_amd64.s:1371
</pre></code>

Based on these tests, it doesn't seem to be a compatibility issue between Hyper-V and prysm. This error seems to mean that the previous block information cannot be read from the DB. However, if I request the previous block or a block in the past through the API, I can see the value in the browser without any problem. (e.g. http://127.0.0.1:3500/eth/v2/beacon/blocks/3058411)

Is there any way I can check if the DB is really corrupted? I've conducted resync 2 times during Jan 2022. (Sync was performed on different machines, respectively.) It doesn't make sense to me that the DB continues to be corrupted though.

**Comment by prestonvanloon** [comment]
> Is there any way I can check if the DB is really corrupted?

Not really. We would likely write some kind of one-off tool to query the database to check whether or not the block actually exists in the database. If it does exist, we can use that database to try to step through the requests to find out why you are receiving the error message about pre-state.

**Comment by e6a6** [comment]
> Any chance you could upload the beaconchain.db? We could inspect to see if there is some corrupt data there.

Because the size of beaconchain.db is too big, it takes time to prepare for upload after compression. Please allow some time. I will post the link after uploading it to my NAS. (maybe 10~12 hours later??)

**Comment by nisdas** [comment]
Ok i just saw from the flags you used that you were running with the slasher. That seems like it would potentially cause issues, please remove the slasher flag and run your node again.

**Comment by e6a6** [comment]
I shut down the beacon and validator running on machine A (native widnows 10). And I ran prysm on Ubuntu VM (Hyper-V) on Machine B again. This time I removed the slasher flag. No more errors while running node for 2 hours. Removing the slasher flag seems to solve the problem. However, it is a temporary expedient and not a fundamental solution. The causal relationship between slasher and "could not find block in DB" error is still unclear.

Here is the link to the beaconchain.db file I uploaded to my NAS. It's around 33 GB, and 56 GB when uncompressed.
http://gofile.me/3MwWE/qu0kSICFh

slasher.db is also thought to be helpful in identifying the cause. My slasher.db file is also ready to upload, may I post a download link here? If I publish slasher.db here could it be a threat to my security (wallet, personal information, etc...)?

**Comment by nisdas** [comment]
I dont think there is anything personally identifiable in the slasher db, but I will let @rauljordan answer that. 

> Removing the slasher flag seems to solve the problem. However, it is a temporary expedient and not a fundamental solution. 

So I think the error that you have seen is from the slasher, in this case the correlation is quite strong. There have been earlier reports also of that particular error log being triggered by the slasher. Without the slasher flag, I don't expect that you will see that error anymore. The other issue with the `cached trie being invalid` seems to be related to #10120, all in all I dont think there is anything wrong with your hardware (A or B) . Let us know if you continue facing any other issues for your node in block proposals.

**Comment by e6a6** [comment]
Okay. When the block is proposed, i will add comment. According to my node's past experience, the next block proposal is likely to occur within a month. And, I don't know why slasher is giving me this error only for me. Others seem to have no problem using the slasher. Is there anything I can do to help in figuring out the slasher error?

**Comment by nisdas** [comment]
I am unsure on the reason why the slasher gives you the error, @rauljordan any ideas why ?

**Comment by e6a6** [comment]
After removing the slasher flag, the node works without any errors for several days. However, it is not clear why a DB error occurs when the slasher is enabled.

In the meantime I did another test. I wanted to check if the use of a virtual machine (Hyper-V) really caused problems with slasher and the DB. So I ran prysm 2.0.6 on native windows 10 pro, not a virtual machine. All existing DBs were deleted and sync was performed from the beginning in the native windows environment.

Sync was completed in about two days, and I started the validator with slasher flag enabled. But after a while I met "could not find block in DB" error again.

So, this error is not caused by the virtual machine. Because it is the same symptom in the native environment. I removed the slasher flagg and started the validator again. No error occurs. Same as for Ubuntu VM. It is presumed that an unknown bug is inherent in the slasher.

**Comment by e6a6** [comment]
A few hours ago, my node successfully proposed a block. Removing the slasher flag from the beacon helped fix the problem. Just before the block proposal, there was no error in the validator, but the following warning in the beacon.

<pre><code>
Feb 10 23:28:59 ETH2-VM prysm.sh[904]: time="2022-02-10 23:28:59" level=warning msg="Cached deposit trie is invalid, rebuilding it now: wanted the canonical count of 295450 but received 295081" prefix="rpc/validator"
</pre></code>

This warning was the same occurred before the block proposal, even before removing the slasher flag. The solution of this warning is unknown, but I think there is no major problem, probably. However, It is still a big problem that slashers must be disabled to resolve error(could not find block in DB). I tested the lighthouse v2.1.2, another client that supports slasher, to see if this error is client dependent. I used the same hardware, Machine A, to finish syncing and run the validator. In lighthouse, enabling the slasher didn't give me a similar error every epoch.
